### PR TITLE
qa: psalm improvements

### PR DESCRIPTION
**Pull request by gsteel**
|    Q          |   A
|-------------- | ------
| Bugfix        | yes
| BC Break      | no
| QA            | yes

### Description

Trying to make some improvements to `input-filter` has led me down a psalm rabbit hole…

Whilst this pull is a general reduction in the baseline and improvement for AbstactValidator and the plugin manager, it has introduced some new errors that I've baselined because I don't have the time - most of them appear to be constructor incompatibilities in the many validators - i.e. annotating each individual constructor with `@param array<string, mixed>|Traversable<string, mixed> $options`

Spotted a couple of void returns when they were declared as whatever|null

I'll do a review and point out as much as possible - Hopefully, on balance this is an improvement…

### CODE REVIEW

File: src/AbstractValidator.php [hunk 1]
```diff
@@ -26,6 +26,20 @@
...
+ *     messageTemplates: array<string, string>,
+ *     messageVariables: array<string, mixed>,
+ *     translator: Translator\TranslatorInterface|null,
+ *     translatorTextDomain: string|null,
+ *     translatorEnabled: bool,
+ *     valueObscured: bool,
+ * }
+ * @property array<string, mixed> $options
+ * @property array<string, string> $messageTemplates
+ * @property array<string, mixed> $messageVariables
```
Comment (gsteel) [inline src/AbstractValidator.php:40]: declaring these magic properties has caused an increase in psalm errors that should be easily fixed in individual validators

File: src/AbstractValidator.php [hunk 2]
```diff
@@ -296,29 +322,29 @@ public function __get($property)
     protected function createMessage($messageKey, $value)
     {
         if (! isset($this->abstractOptions['messageTemplates'][$messageKey])) {
-            return;
+            return null;
```
Comment (gsteel) [inline src/AbstractValidator.php:323]: This was a bug right?

File: src/AbstractValidator.php [hunk 3]
```diff
@@ -351,11 +377,19 @@ protected function error($messageKey, $value = null)
...
             $value = $this->value;
         }
 
-        $this->abstractOptions['messages'][$messageKey] = $this->createMessage($messageKey, $value);
+        $message = $this->createMessage($messageKey, $value);
+        if (! is_string($message)) {
+            return;
+        }
+
+        $this->abstractOptions['messages'][$messageKey] = $message;
```
Comment (gsteel) [inline src/AbstractValidator.php:388]: Message is possibly null here, so an early return prevents `null` from being added to the list of messages - perhaps a subtle BC break?

File: src/AbstractValidator.php [hunk 4]
```diff
@@ -427,7 +462,7 @@ public function setTranslator(?Translator\TranslatorInterface $translator = null
     public function getTranslator()
     {
         if (! $this->isTranslatorEnabled()) {
-            return;
+            return null;
```
Comment (gsteel) [inline src/AbstractValidator.php:461]: Another void return which _should_ be null

File: src/ValidatorChain.php [hunk 5]
```diff
@@ -239,12 +244,13 @@ public function isValid($value, $context = null)
...
         foreach ($this->validators as $element) {
             $validator = $element['instance'];
+            assert($validator instanceof ValidatorInterface);
             if ($validator->isValid($value, $context)) {
                 continue;
             }
             $result         = false;
             $messages       = $validator->getMessages();
-            $this->messages = array_replace_recursive($this->messages, $messages);
+            $this->messages = array_replace($this->messages, $messages);
```
Comment (gsteel) [inline src/ValidatorChain.php:253]: messages is `array<string, string>` therefore `array_replace` is sufficient and retains the existing type for psalm

File: src/AbstractValidator.php [hunk 6]
```diff
@@ -26,6 +26,20 @@
...
+ * @psalm-type AbstractOptions = array{
+ *     messages: array<string, string>,
+ *     messageTemplates: array<string, string>,
+ *     messageVariables: array<string, mixed>,
+ *     translator: Translator\TranslatorInterface|null,
+ *     translatorTextDomain: string|null,
+ *     translatorEnabled: bool,
+ *     valueObscured: bool,
+ * }
+ * @property array<string, mixed> $options
```
Comment (Ocramius) [inline src/AbstractValidator.php:38]: Are these `public`? Because `@property` makes them `public` :thinking:

File: src/AbstractValidator.php [hunk 7]
```diff
@@ -78,20 +91,23 @@ abstract class AbstractValidator implements
...
      *
-     * @param array|Traversable $options
+     * @param array<string, mixed>|Traversable<string, mixed> $options
      */
     public function __construct($options = null)
     {
         // The abstract constructor allows no scalar values
         if ($options instanceof Traversable) {
+            /** @psalm-var array<string, mixed> $options */
             $options = ArrayUtils::iteratorToArray($options);
```
Comment (Ocramius) [inline src/AbstractValidator.php:99]: TBH, sounds like we could improve inference here, in future

File: src/AbstractValidator.php (hunk 2 above)
Comment (froschdesign) [reply src/AbstractValidator.php:323]: Very old remains.

File: src/AbstractValidator.php (hunk 4 above)
Comment (froschdesign) [reply src/AbstractValidator.php:461]: Same here.

File: src/AbstractValidator.php (hunk 7 above)
Comment (gsteel) [reply src/AbstractValidator.php:99]: I'll send in a pull for `ArrayUtils::iteratorToArray` asap

File: src/AbstractValidator.php (hunk 6 above)
Comment (gsteel) [reply src/AbstractValidator.php:38]: I guess if they're undeclared they are already public no? Typically they'd be protected in extending classes, if set at all, so I guess not. I'm unsure as to why `$abstractOptions` exists at all and why these properties aren't just declared on the abstract

All I actually wanted to achieve was to nail down what `getMessages()` returns but doing that is a bit of an endless battle.

File: src/AbstractValidator.php (hunk 7 above)
Comment (gsteel) [reply src/AbstractValidator.php:99]: https://github.com/laminas/laminas-stdlib/pull/62

Comment (Ocramius) [reply src/AbstractValidator.php:99]: @gsteel released!

File: .psr-container-stub.php [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<?php
```
Comment (Ocramius) [inline .psr-container-stub.php]: Following discussion at https://github.com/Lctrs/psalm-psr-container-plugin/issues/545#issuecomment-1150465946, can we:

 1. export-ignore this
 2. perhaps rename it to `.psr-container-stub.php.stub`?

Comment (Ocramius) [reply .psr-container-stub.php]: We would probably drop a line of comment explaining why this doesn't end with `.php` :D

File: src/AbstractValidator.php (hunk 6 above)
Comment (Ocramius) [reply src/AbstractValidator.php:38]: Let's keep it as-is then: still a good improvement :+1:

File: src/AbstractValidator.php (hunk 3 above)
Comment (Ocramius) [reply src/AbstractValidator.php:388]: I think it's OK to prevent `null` from leaking in here.

**Review by Ocramius — APPROVED** [review]
LGTM! Waiting for @gsteel's feedback and potential changes on the stub file, before merging

File: .psr-container-stub.php (hunk 8 above)
Comment (gsteel) [reply .psr-container-stub.php]: 👍 - Had a general tidy-up of the dot-files/ignores etc too and got rid of coveralls

**Comment by Ocramius** [comment]
Thanks @gsteel!

File: .psr-container-stub.php (hunk 8 above)
Comment (Ocramius) [reply .psr-container-stub.php]: Indeed, awesome!

**Pull request merged**